src/app/clusters/network-commissioning/network-commissioning.h

```diff
@@ -18,23 +18,105 @@
-                                                       ByteSpan operationalDataset, uint64_t breadcrumb, uint32_t timeoutMs);
-void OnAddOrUpdateWiFiNetworkCommandCallbackInternal(app::CommandHandler *, const app::ConcreteCommandPath & commandPath,
-                                                     ByteSpan ssid, ByteSpan credentials, uint64_t breadcrumb, uint32_t timeoutMs);
-void OnConnectNetworkCommandCallbackInternal(app::CommandHandler *, const app::ConcreteCommandPath & commandPath,
-                                             ByteSpan networkID, uint64_t breadcrumb, uint32_t timeoutMs);
-} // namespace NetworkCommissioning
 
+// TODO: Use macro to disable some wifi or thread
+class Instance : public CommandHandlerInterface,
+                 public AttributeAccessInterface,
+                 public DeviceLayer::NetworkCommissioning::Internal::WirelessDriver::ConnectCallback,
+                 public DeviceLayer::NetworkCommissioning::WiFiDriver::ScanCallback,
+                 public DeviceLayer::NetworkCommissioning::ThreadDriver::ScanCallback
+{
+public:
+    /**
+     * Register will register the network commissioning instance to the attribute and command dispatching route.
+     */
+    CHIP_ERROR Register();
+
+    // CommandHandlerInterface
+    void InvokeCommand(HandlerContext & ctx) override;
+
+    // AttributeAccessInterface
+    CHIP_ERROR Read(const ConcreteReadAttributePath & aPath, AttributeValueEncoder & aEncoder) override;
+    CHIP_ERROR Write(const ConcreteDataAttributePath & aPath, AttributeValueDecoder & aDecoder) override;
+
+    // WirelessDriver::ConnectCallback
+    void OnResult(DeviceLayer::NetworkCommissioning::Status commissioningError, CharSpan errorText,
+                  int32_t interfaceStatus) override;
+
+    // WiFiDriver::ScanCallback
+    void OnFinished(DeviceLayer::NetworkCommissioning::Status err, CharSpan debugText,
+                    DeviceLayer::NetworkCommissioning::WiFiScanResponseIterator * networks) override;
+
+    // ThreadDriver::ScanCallback
+    void OnFinished(DeviceLayer::NetworkCommissioning::Status err, CharSpan debugText,
+                    DeviceLayer::NetworkCommissioning::ThreadScanResponseIterator * networks) override;
+
+private:
+    static constexpr uint32_t kFeatureMapWiFi     = 0b0000'0001;
+    static constexpr uint32_t kFeatureMapThread   = 0b0000'0010;
+    static constexpr uint32_t kFeatureMapEthernet = 0b0000'0100;
+
+    static constexpr uint32_t kFeatureMapWireless = 0b0000'0011;
+
+    static void _OnCommissioningComplete(const DeviceLayer::ChipDeviceEvent * event, intptr_t arg);
+    void OnCommissioningComplete(CHIP_ERROR err);
+
+    const uint32_t mType;
+
+    DeviceLayer::NetworkCommissioning::Internal::WirelessDriver * mpWirelessDriver;
+    DeviceLayer::NetworkCommissioning::Internal::BaseDriver * mpBaseDriver;
+
+    // TODO: Use std::variant
```

Comment:
std::variant is not available in c++14.  However, we have our own at src/lib/support/Variant.h.  I've used it extensively.  It works well.